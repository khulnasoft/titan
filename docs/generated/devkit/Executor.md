# Type alias: Executor\<T\>

Ƭ **Executor**\<`T`\>: (`options`: `T`, `context`: [`ExecutorContext`](../../devkit/documents/ExecutorContext)) => `Promise`\<\{ `success`: `boolean` }\> \| `AsyncIterableIterator`\<\{ `success`: `boolean` }\>

Implementation of a target of a project

#### Type parameters

| Name | Type  |
| :--- | :---- |
| `T`  | `any` |

#### Type declaration

▸ (`options`, `context`): `Promise`\<\{ `success`: `boolean` }\> \| `AsyncIterableIterator`\<\{ `success`: `boolean` }\>

##### Parameters

| Name      | Type                                                        |
| :-------- | :---------------------------------------------------------- |
| `options` | `T`                                                         |
| `context` | [`ExecutorContext`](../../devkit/documents/ExecutorContext) |

##### Returns

`Promise`\<\{ `success`: `boolean` }\> \| `AsyncIterableIterator`\<\{ `success`: `boolean` }\>
