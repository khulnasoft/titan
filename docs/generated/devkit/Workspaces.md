# Class: Workspaces

**`Deprecated`**

This will be removed in v19. Use [readProjectsConfigurationFromProjectGraph](../../devkit/documents/readProjectsConfigurationFromProjectGraph) instead.

## Table of contents

### Constructors

- [constructor](../../devkit/documents/Workspaces#constructor)

### Methods

- [readWorkspaceConfiguration](../../devkit/documents/Workspaces#readworkspaceconfiguration)

## Constructors

### constructor

• **new Workspaces**(`root`): [`Workspaces`](../../devkit/documents/Workspaces)

#### Parameters

| Name   | Type     |
| :----- | :------- |
| `root` | `string` |

#### Returns

[`Workspaces`](../../devkit/documents/Workspaces)

## Methods

### readWorkspaceConfiguration

▸ **readWorkspaceConfiguration**(): [`ProjectsConfigurations`](../../devkit/documents/ProjectsConfigurations) & [`NxJsonConfiguration`](../../devkit/documents/NxJsonConfiguration)\<`string`[] \| `"*"`\>

#### Returns

[`ProjectsConfigurations`](../../devkit/documents/ProjectsConfigurations) & [`NxJsonConfiguration`](../../devkit/documents/NxJsonConfiguration)\<`string`[] \| `"*"`\>

**`Deprecated`**

Use [readProjectsConfigurationFromProjectGraph](../../devkit/documents/readProjectsConfigurationFromProjectGraph) instead.
