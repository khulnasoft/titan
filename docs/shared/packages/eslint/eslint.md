The ESLint plugin integrates [ESLint](https://eslint.org/) with Nx. It allows you to run ESLint through Nx with caching enabled. It also includes code generators to help you set up ESLint in your workspace.

## Setting Up @titan/eslint

### Installation

{% callout type="note" title="Keep Nx Package Versions In Sync" %}
Make sure to install the `@titan/eslint` version that matches the version of `nx` in your repository. If the version numbers get out of sync, you can encounter some difficult to debug errors. You can [fix Nx version mismatches with this recipe](/recipes/tips-n-tricks/keep-nx-versions-in-sync).
{% /callout %}

In any Nx workspace, you can install `@titan/eslint` by running the following command:

{% tabs %}
{% tab label="Nx 18+" %}

```shell {% skipRescope=true %}
nx add @titan/eslint
```

This will install the correct version of `@titan/eslint`.

### How @titan/eslint Infers Tasks

The `@titan/eslint` plugin will create a task for any project that has an ESLint configuration file present. Any of the following files will be recognized as an ESLint configuration file:

- `.eslintrc`
- `.eslintrc.js`
- `.eslintrc.cjs`
- `.eslintrc.yaml`
- `.eslintrc.yml`
- `.eslintrc.json`
- `eslint.config.js`

Because ESLint applies configuration files to all subdirectories, the `@titan/eslint` plugin will also infer tasks for projects in subdirectories. So, if there is an ESLint configuration file in the root of the repository, every project will have an inferred ESLint task.

### View Inferred Tasks

To view inferred tasks for a project, open the [project details view](/concepts/inferred-tasks) in Nx Console or run `nx show project my-project --web` in the command line.

### @titan/eslint Configuration

The `@titan/eslint/plugin` is configured in the `plugins` array in `nx.json`.

```json {% fileName="nx.json" %}
{
  "plugins": [
    {
      "plugin": "@titan/eslint/plugin",
      "options": {
        "targetName": "lint"
      }
    }
  ]
}
```

- The `targetName` option controls the name of the inferred ESLint tasks. The default name is `lint`.

{% /tab %}
{% tab label="Nx 17" %}

Install the `@titan/eslint` package with your package manager.

```shell {% skipRescope=true %}
npm add -D @titan/eslint
```

{% /tab %}
{% tab label="Nx < 17" %}

Install the `@nx/linter` package with your package manager.

```shell
npm add -D @nx/linter
```

{% /tab %}
{% /tabs %}

## Lint

You can lint an application or a library with the following command:

```shell
nx lint my-project
```

## Utils

- [convert-to-flat-config](/nx-api/eslint/generators/convert-to-flat-config) - Converts the workspace's [ESLint](https://eslint.org/) configs to the new [Flat Config](https://eslint.org/blog/2022/08/new-config-system-part-2)

## ESLint plugin

Read about our dedicated ESLint plugin - [eslint-plugin-nx](/nx-api/eslint-plugin/documents/overview).
