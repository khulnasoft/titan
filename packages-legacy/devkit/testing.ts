export * from '@titan/devkit/testing';
