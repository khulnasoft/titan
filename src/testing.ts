export {cold, hot} from 'jasmine-marbles';
export {readAll} from './utils/testing';
