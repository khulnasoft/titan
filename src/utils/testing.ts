import {Observable} from 'rxjs/Observable';
import {toArray} from 'rxjs/operator/toArray';
import {toPromise} from 'rxjs/operator/toPromise';

export function readAll<T>(o: Observable<T>): Promise<T[]> {
  return toPromise.call(toArray.call(o));
}
