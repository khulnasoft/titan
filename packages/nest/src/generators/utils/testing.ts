import { addProjectConfiguration, Tree } from '@titan/devkit';
import { createTreeWithEmptyWorkspace } from '@titan/devkit/testing';

export function createTreeWithNestApplication(appName: string): Tree {
  const tree = createTreeWithEmptyWorkspace();
  addProjectConfiguration(tree, appName, {
    root: `${appName}`,
    sourceRoot: `${appName}/src`,
    projectType: 'application',
    targets: {},
  });
  return tree;
}
