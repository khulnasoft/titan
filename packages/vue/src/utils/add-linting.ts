import { Tree } from 'nx/src/generators/tree';
import { Linter, lintProjectGenerator } from '@titan/eslint';
import { joinPathFragments } from 'nx/src/utils/path';
import {
  addDependenciesToPackageJson,
  runTasksInSerial,
  updateJson,
} from '@titan/devkit';
import { extraEslintDependencies } from './lint';
import {
  addExtendsToLintConfig,
  isEslintConfigSupported,
} from '@titan/eslint/src/generators/utils/eslint-file';

export async function addLinting(
  host: Tree,
  options: {
    linter: Linter;
    name: string;
    projectRoot: string;
    unitTestRunner?: 'vitest' | 'none';
    setParserOptionsProject?: boolean;
    skipPackageJson?: boolean;
    rootProject?: boolean;
    addPlugin?: boolean;
  },
  projectType: 'lib' | 'app'
) {
  if (options.linter === Linter.EsLint) {
    const lintTask = await lintProjectGenerator(host, {
      linter: options.linter,
      project: options.name,
      tsConfigPaths: [
        joinPathFragments(options.projectRoot, `tsconfig.${projectType}.json`),
      ],
      unitTestRunner: options.unitTestRunner,
      skipFormat: true,
      setParserOptionsProject: options.setParserOptionsProject,
      rootProject: options.rootProject,
      addPlugin: options.addPlugin,
    });

    if (isEslintConfigSupported(host)) {
      addExtendsToLintConfig(host, options.projectRoot, [
        'plugin:vue/vue3-essential',
        'eslint:recommended',
        '@vue/eslint-config-typescript',
        '@vue/eslint-config-prettier/skip-formatting',
      ]);
    }

    editEslintConfigFiles(host, options.projectRoot, options.rootProject);

    let installTask = () => {};
    if (!options.skipPackageJson) {
      installTask = addDependenciesToPackageJson(
        host,
        extraEslintDependencies.dependencies,
        extraEslintDependencies.devDependencies
      );
    }

    return runTasksInSerial(lintTask, installTask);
  } else {
    return () => {};
  }
}

export function editEslintConfigFiles(
  tree: Tree,
  projectRoot: string,
  rootProject?: boolean
) {
  if (tree.exists(joinPathFragments(projectRoot, 'eslint.config.js'))) {
    const fileName = joinPathFragments(projectRoot, 'eslint.config.js');
    updateJson(tree, fileName, (json) => {
      let updated = false;
      for (let override of json.overrides) {
        if (override.parserOptions) {
          if (!override.files.includes('*.vue')) {
            override.files.push('*.vue');
          }
          updated = true;
        }
      }
      if (!updated) {
        json.overrides = [
          {
            files: ['*.ts', '*.tsx', '*.js', '*.jsx', '*.vue'],
            rules: { 'vue/multi-word-component-names': 'off' },
          },
        ];
      }
      return json;
    });
  } else {
    const fileName = joinPathFragments(projectRoot, '.eslintrc.json');
    updateJson(tree, fileName, (json) => {
      let updated = false;
      for (let override of json.overrides) {
        if (override.parserOptions) {
          if (!override.files.includes('*.vue')) {
            override.files.push('*.vue');
          }
          updated = true;
        }
      }
      if (!updated) {
        json.overrides = [
          {
            files: ['*.ts', '*.tsx', '*.js', '*.jsx', '*.vue'],
            rules: { 'vue/multi-word-component-names': 'off' },
          },
        ];
      }
      return json;
    });
  }

  // Edit root config too
  if (tree.exists('.eslintrc.base.json')) {
    updateJson(tree, '.eslintrc.base.json', (json) => {
      for (let override of json.overrides) {
        if (
          override.rules &&
          '@nx/enforce-module-boundaries' in override.rules
        ) {
          if (!override.files.includes('*.vue')) {
            override.files.push('*.vue');
          }
        }
      }
      return json;
    });
  } else if (tree.exists('.eslintrc.json') && !rootProject) {
    updateJson(tree, '.eslintrc.json', (json) => {
      for (let override of json.overrides) {
        if (
          override.rules &&
          '@nx/enforce-module-boundaries' in override.rules
        ) {
          if (!override.files.includes('*.vue')) {
            override.files.push('*.vue');
          }
        }
      }
      return json;
    });
  }
}
