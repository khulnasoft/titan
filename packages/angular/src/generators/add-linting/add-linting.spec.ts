import {
  ProjectConfiguration,
  Tree,
  addProjectConfiguration,
  readJson,
} from '@titan/devkit';
import { createTreeWithEmptyWorkspace } from '@titan/devkit/testing';
import * as linter from '@titan/eslint';
import { addLintingGenerator } from './add-linting';

describe('addLinting generator', () => {
  let tree: Tree;
  const appProjectName = 'ng-app1';
  const appProjectRoot = `apps/${appProjectName}`;

  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace({ layout: 'apps-libs' });

    addProjectConfiguration(tree, appProjectName, {
      root: appProjectRoot,
      prefix: 'myOrg',
      projectType: 'application',
      targets: {},
    } as ProjectConfiguration);
  });

  it('should invoke the lintProjectGenerator', async () => {
    jest.spyOn(linter, 'lintProjectGenerator');

    await addLintingGenerator(tree, {
      prefix: 'myOrg',
      projectName: appProjectName,
      projectRoot: appProjectRoot,
      skipFormat: true,
    });

    expect(linter.lintProjectGenerator).toHaveBeenCalled();
  });

  it('should add the Angular specific EsLint devDependencies', async () => {
    await addLintingGenerator(tree, {
      prefix: 'myOrg',
      projectName: appProjectName,
      projectRoot: appProjectRoot,
      skipFormat: true,
    });

    const { devDependencies } = readJson(tree, 'package.json');
    expect(devDependencies['@angular-eslint/eslint-plugin']).toBeDefined();
    expect(
      devDependencies['@angular-eslint/eslint-plugin-template']
    ).toBeDefined();
    expect(devDependencies['@angular-eslint/template-parser']).toBeDefined();
  });

  it('should correctly generate the .eslintrc.json file', async () => {
    await addLintingGenerator(tree, {
      prefix: 'myOrg',
      projectName: appProjectName,
      projectRoot: appProjectRoot,
      skipFormat: true,
    });

    const eslintConfig = readJson(tree, `${appProjectRoot}/.eslintrc.json`);
    expect(eslintConfig).toMatchSnapshot();
  });
});
