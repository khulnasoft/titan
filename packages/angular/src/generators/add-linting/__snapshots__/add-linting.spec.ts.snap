// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addLinting generator should correctly generate the .eslintrc.json file 1`] = `
{
  "extends": [
    "../../.eslintrc.json",
  ],
  "ignorePatterns": [
    "!**/*",
  ],
  "overrides": [
    {
      "extends": [
        "plugin:@titan/angular",
        "plugin:@angular-eslint/template/process-inline-templates",
      ],
      "files": [
        "*.ts",
      ],
      "rules": {
        "@angular-eslint/component-selector": [
          "error",
          {
            "prefix": "my-org",
            "style": "kebab-case",
            "type": "element",
          },
        ],
        "@angular-eslint/directive-selector": [
          "error",
          {
            "prefix": "myOrg",
            "style": "camelCase",
            "type": "attribute",
          },
        ],
      },
    },
    {
      "extends": [
        "plugin:@titan/angular-template",
      ],
      "files": [
        "*.html",
      ],
      "rules": {},
    },
  ],
}
`;
